peps/pep-0750.rst

```diff
@@ -0,0 +1,813 @@
+            chunk = super().__new__(cls, decoded)
+            chunk._raw = raw
+            return chunk
+
+        @property
+        def raw(self):
+            return self._raw
+
+Interpolation
+-------------
+
+An ``Interpolation`` is the data structure representing an expression inside the tag
+string. Interpolations enable a delayed evaluation model, where the interpolation
+expression is computed, transformed, memoized, or processed in any way.
+
+In addition, the original text of the interpolation expression is made available to the
+tag function. This can be useful for debugging or metaprogramming.
+
+``Interpolation`` is a ``Protocol`` which will be made available from ``typing``. It
+has the following definition:
+
+.. code-block:: python
+
+    @runtime_checkable
+    class Interpolation(Protocol):
+        def __len__(self):
+            ...
+
+        def __getitem__(self, index: int):
+            ...
+
+        def getvalue(self) -> Callable[[], Any]:
+            ...
+
+        expr: str
+        conv: Literal["a", "r", "s"] | None
+        format_spec: str | None
+
+Given this example interpolation:
+
+.. code-block:: python
+
+    mytag'{trade!r:some-formatspec}'
+
+these attributes are as follows:
+
+* ``getvalue`` is a zero argument closure for the interpolation. In this case, ``lambda: trade``.
+
+* ``expr`` is the *expression text* of the interpolation. Example: ``'trade'``.
+
+* ``conv`` is the
+  `optional conversion <https://docs.python.org/3/library/string.html#format-string-syntax>`_
+  to be used by the tag function, one of ``r``, ``s``, and ``a``, corresponding to repr, str,
+  and ascii conversions. Note that as with f-strings, no other conversions are supported.
+  Example: ``'r'``.
+
+* ``format_spec`` is the optional `format_spec string <https://docs.python.org/3/library/string.html#format-specification-mini-language>`_.
+  A format_spec is eagerly evaluated if it contains any expressions before being passed to the tag
+  function. Example: ``'some-formatspec'``.
+
+In all cases, the tag function determines what to do with valid ``Interpolation``
+attributes.
+
+In the CPython reference implementation, implementing ``Interpolation`` in C would
+use the equivalent `Struct Sequence Objects
+<https://docs.python.org/3/c-api/tuple.html#struct-sequence-objects>`_ (see
+such code as `os.stat_result
+<https://docs.python.org/3/library/os.html#os.stat_result>`_). For purposes of this
+PEP, here is an example of a pure Python implementation:
+
+.. code-block:: python
+
+    class InterpolationConcrete(NamedTuple):
+        getvalue: Callable[[], Any]
+        expr: str
+        conv: Literal['a', 'r', 's'] | None = None
+        format_spec: str | None = None
+
+Interpolation Expression Evaluation
+-----------------------------------
+
+Expression evaluation for interpolations is the same as in :pep:`498#expression-evaluation`,
+except that all expressions are always implicitly wrapped with a ``lambda``:
```

Comment:
You may want to special-case class scopes, like in PEP 649 and 695.

Reply:
@JelleZijlstra not certain what you mean here. So let me try to consider how those PEPs might be related.

* PEP 649 is with respect to deferred annotations, but in PEP 750 we really just introduce a fairly standard lambda wrapper - if implicitly - in constructing `getvalue` in the interpolation. This wrapping is with respect to the lexical scope, just like f-strings.
* PEP 695 is with respect to typing (so maybe it's possible to refine the protocols for  `Interpolation` and `Decoded`?).

Reply:
I am referring to the behavior in class scopes. With your proposal, this will not work as expected, because lambdas in class scopes can't see the enclosing class scope:
```
class Cls:
    name = "World"
    value = greet"Hello {name}"
```

PEP 649 and 695 use a new mechanism to create scopes nested in classes that do have access to the enclosing scope.

Reply:
I see - this would be what is decribed as annotation scopes: https://docs.python.org/3/reference/executionmodel.html#annotation-scopes
and in https://peps.python.org/pep-0695/#scoping-behavior

I did see that f-strings resolves `name`, presumably because it is immediate - no need to look at this scoping behavior. Interesting.

Reply:
@davepeck put together an initial PR on @lysnikolaou 's ref implementation to add class scoping to the lambdas created for `Interpolation` objects: https://github.com/lysnikolaou/cpython/pull/44

We plan to further investigate on well this can work to support tag strings, namely the HTML component example I mentioned in my previous comment.